common/djangoapps/util/keyword_substitution.py

```diff
@@ -48,35 +41,37 @@ def substitute_keywords(string, user=None, course=None):
     them by calling the corresponding functions stored in KEYWORD_FUNCTION_MAP.
 
     Functions stored in KEYWORD_FUNCTION_MAP must return a replacement string.
-    Also, functions imported from other modules must be wrapped in a
-    new function if they don't take in user_id and course_id. This simplifies
-    the loop below, and reduces the need for piling up if elif else statements
-    when the keyword pool grows.
     """
-    if user is None or course is None:
-        # Cannot proceed without course and user information
-        return string
 
-    for key, func in KEYWORD_FUNCTION_MAP.iteritems():
+    # do this lazily to avoid unneeded database hits
+    KEYWORD_FUNCTION_MAP = {
+        '%%USER_ID%%': lambda: anonymous_id_from_user_id(user_id),
+        '%%USER_FULLNAME%%': lambda: context.get('name'),
+        '%%COURSE_DISPLAY_NAME%%': lambda: context.get('course_title'),
+        '%%COURSE_END_DATE%%': lambda: context.get('course_end_date'),
+    }
+
+    for key in KEYWORD_FUNCTION_MAP.keys():
         if key in string:
-            substitutor = func(user, course)
-            string = string.replace(key, substitutor)
+            substitutor = KEYWORD_FUNCTION_MAP[key]
+            string = string.replace(key, substitutor())
 
     return string
 
 
-def substitute_keywords_with_data(string, user_id=None, course_id=None):
+def substitute_keywords_with_data(string, context):
     """
-    Given user and course ids, replaces all %%-encoded words in the given string
+    Given an email context, replaces all %%-encoded words in the given string
+    `context` is a dictionary that should include `user_id` and `course_title`
+    keys
     """
 
     # Do not proceed without parameters: Compatibility check with existing tests
     # that do not supply these parameters
-    if user_id is None or course_id is None:
-        return string
+    user_id = context.get('user_id')
+    course_title = context.get('course_title')
 
-    # Grab user objects
-    user = User.objects.get(id=user_id)
-    course = modulestore().get_course(course_id, depth=0)
+    if user_id is None or course_title is None:
```

Comment:
It's strange that we only check for course_title and user_id (since name and course_end_date may also be referenced), but I see that is how it was before.